internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+	defer bdb.metaMtx.RUnlock()
+	info := blockchain.Info[H, N]{
+		BestHash:        bdb.meta.BestHash,
+		BestNumber:      bdb.meta.BestNumber,
+		GenesisHash:     bdb.meta.GenesisHash,
+		FinalizedHash:   bdb.meta.FinalizedHash,
+		FinalizedNumber: bdb.meta.FinalizedNumber,
+		FinalizedState: &struct {
+			Hash   H
+			Number N
+		}{bdb.meta.FinalizedState.Hash, bdb.meta.FinalizedState.Number},
+		NumberLeaves: bdb.leaves.Count(),
+		BlockGap:     bdb.meta.BlockGap,
+	}
+	return info
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Status(hash H) (blockchain.BlockStatus, error) {
+	header, err := bdb.Header(hash)
+	if err != nil {
+		return 0, err
+	}
+	if header != nil {
+		return blockchain.BlockStatusInChain, nil
+	}
+	return blockchain.BlockStatusUnknown, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Number(hash H) (*N, error) {
+	meta, err := bdb.HeaderMetadata(hash)
+	if err != nil {
+		return nil, err
+	}
+	return &meta.Number, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Hash(number N) (*H, error) {
+	header, err := readHeader[H, N, Header](
+		bdb.db,
+		columns.KeyLookup,
+		columns.Header,
+		generic.BlockIDNumber[N]{Inner: number},
+	)
+	if err != nil {
+		return nil, err
+	}
+	if header == nil {
+		return nil, nil
+	}
+	h := (*header).Hash()
+	return &h, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) BlockHashFromID(id generic.BlockID) (*H, error) {
+	switch id := id.(type) {
+	case generic.BlockIDHash[H]:
+		return &id.Inner, nil
+	case generic.BlockIDNumber[N]:
+		return bdb.Hash(id.Inner)
+	default:
+		panic("unsupported block id type")
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) BlockNumberFromID(id generic.BlockID) (*N, error) {
+	switch id := id.(type) {
+	case generic.BlockIDHash[H]:
+		return bdb.Number(id.Inner)
+	case generic.BlockIDNumber[N]:
+		return &id.Inner, nil
+	default:
+		panic("unsupported block id type")
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Body(hash H) ([]runtime.Extrinsic, error) {
+	bdb.pinnedBlocksCacheMtx.RLock()
+	defer bdb.pinnedBlocksCacheMtx.RUnlock()
+	body := bdb.pinnedBlocksCache.Body(hash)
+	if body != nil {
+		return *body, nil
+	}
+
+	return bdb.bodyUncached(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Justifications(hash H) (runtime.Justifications, error) {
+	bdb.pinnedBlocksCacheMtx.RLock()
+	defer bdb.pinnedBlocksCacheMtx.RUnlock()
+	justifications := bdb.pinnedBlocksCache.Justifications(hash)
+	if justifications != nil {
+		return justifications, nil
+	}
+
+	return bdb.justificationsUncached(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) LastFinalized() (H, error) {
+	bdb.metaMtx.RLock()
+	defer bdb.metaMtx.RUnlock()
+	return bdb.meta.FinalizedHash, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Leaves() ([]H, error) {
+	bdb.leavesMtx.RLock()
+	defer bdb.leavesMtx.RUnlock()
+	return bdb.leaves.Hashes(), nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) DisplacedLeavesAfterFinalizing(blockNumber N) ([]H, error) {
```

Comment:
Ill also mention here even though I alluded to it in a separate function, do we need this? What is the use case for getting this info?

Reply:
This function is part of `blockchain.Backend` and will be used in future work.

Reply:
Will resolve this since i commented in a separate function